Support for ternary operator

georgejecook opened the pull request:
- adds ConditionalExpression
- decouples questionmark token from print
- adds parsing for conditional expression
- adds support for ConditionalStatement, which needs to be removed, because we can't support it due to loverloaded equals/assign operator
- adds util method to gather variables under a given syntax branch, which will be used when transpiling conditionals to gather scope

TwitchBronBron commented on .vscode/settings.json:
```diff
@@ -22,5 +22,14 @@
     ],
     "mocha.env": {
         "TS_NODE_TYPE_CHECK": false
+    },
```
Make sure to exclude these once you are finished working on this.

georgejecook replied:
crap. sorry these keep getting in.

TwitchBronBron commented on src/parser/Expression.ts:
```diff
@@ -645,5 +645,112 @@ export class CallfuncExpression implements Expression {
+        let allExpressions = [...testExpressions, ...consequentExpressions, ...alternateExpressions];
+        let allUniqueVarNames = [...new Set([...testUniqueNames, ...consequentUniqueNames, ...alternateUniqueNames])];
+
+        let mutatingExpressions = allExpressions.filter(e => e instanceof CallExpression || e instanceof CallfuncExpression || e instanceof DottedGetExpression);
```
We might need to also include object property references as "expensive" enough to run in the deferred function, because rendezvous could be fairly slow, and the "bslib_simpleTernary" would cause TWO rendezvous.

@chrisdp what do you think?

georgejecook replied:
any way of knowing that up front? I don't think so, no? coz we'd need to know it's a node type. Does that mean any dotteted, get then, irrespective of the presence function calls?

TwitchBronBron replied:
@georgejecook I spoke with @chrisdp and we both agree, there are situations that could cause a rendezvous when accessing object members, so we cannot allow properties to follow the simpleTernary approach since that could be expensive in the long run. So instead we need to restrict the simpleTernary to only variables and literals.

TwitchBronBron replied:
No, there's no way of knowing that up front. So yeah, any dotted get would also need to use the scope-capture implementation.

TwitchBronBron commented on src/parser/Expression.ts:
```diff
@@ -645,5 +645,112 @@ export class CallfuncExpression implements Expression {
+            result = result.concat(this.getScopedFunction(state, this.alternate, alternateUniqueNames));
+            result.push(')');
+        } else {
+            const ternaryCall = `bslib_simpleTernary(${this.test.transpile(state)}, ${this.consequent.transpile(state)}, ${this.alternate.transpile((state))})`;
```
This isn't right. transpile returns `Array<string|SourceNode` so this would end up calling `array.toString()`. You need to break these out.

Something to the effect of
```
result.push('bslib_simpleTernary(');
result.push(...this.test.transpile(state));
result.push(', ');
result.push(...this.consequent.transpile(state));
result.push(', ');
result.push(...this.alternate.transpile(state));
result.push(')');
```

georgejecook replied:
yep - I realized this in my ternary-null coalesce pr and fixed it there. Will fix it here too.

TwitchBronBron commented on src/parser/tests/expression/ConditionalExpression.spec.ts:
```diff
@@ -0,0 +1,266 @@
+    it('throws exception when used in brightscript scope', () => {
+        let { tokens } = Lexer.scan(`a = true ? "human" : "Zombie"`);
+        let { diagnostics } = Parser.parse(tokens, { mode: ParseMode.BrightScript });
+        expect(diagnostics[0]?.code).to.equal(DiagnosticMessages.bsFeatureNotSupportedInBrsFiles('').code);
```
Can you change this to check the message instead? It makes diagnosing errors a little easier since you don't have to go look up the code.

georgejecook replied:
ok

TwitchBronBron commented on src/parser/Expression.ts:
```diff
@@ -645,5 +645,112 @@ export class CallfuncExpression implements Expression {
+        const expressions = getAllExpressions(expression);
+        const varExpressions = expressions.filter(e => e instanceof VariableExpression) as VariableExpression[];
+        const uniqueVarNames = [...new Set(varExpressions.map(e => e.name.text))];
+        return [expressions, varExpressions, uniqueVarNames];
```
I'm not a big fan of using array index locations to represent different values, that's what objects are for. Can you change this to return an object with names?
```
return {
    expressions: expressions,
    variableExpressions: varExpressions,
    variableNames: uniqueVarNames
};
```

Then up in the transpile function you can extract them to local vars in a similar fashion as you were doing:
```
let { expressions, variableExpressions, variableNames } = this.getExpresssionInfo(expr);
```

georgejecook replied:
cool! I was wondering how to do this.

TwitchBronBron commented on src/parser/Expression.ts:
```diff
@@ -645,5 +645,112 @@ export class CallfuncExpression implements Expression {
+ * @param expression
+ * @param expressions
+ */
+function getAllExpressions(expression: Expression, expressions: Expression[] = []): Expression[] {
```
Is there a way we can handle this without needing to hardcode every known expression type? Like perhaps we make each expression EXTEND Expression instead of implement the expression interface, then we can test for Expression in the prototype chain. I'm just worried about adding new expression types over time and not remembering to update this function to account for that.

georgejecook replied:
yes - I put a comment in here saying I think each expression should implement this method. Glad you concur. I'll make it so.

TwitchBronBron commented:
I did a first pass over everything and left some comments.

I know you're not totally finished with the transpile stuff, but we need to make sure to add a bunch of testing for transpile output to make sure it works right. Here's a good example test to use: https://github.com/rokucommunity/brighterscript/blob/master/src/files/BrsFile.spec.ts#L1611

georgejecook commented:
hey @TwitchBronBron I've updated the code quite a bit; still got a cpl more things to do; so wait until I ping in here, in case you see me resolving things. There's a lot of commits here; figured I'd leave it all raw for now, and you can make a call about squashing/rebasing, etc.

I'll be done in 10 or so mins

georgejecook commented:
@TwitchBronBron I think this is ready to go - I just reverted package lock changes; I tried the same with settings; but I don't have any changes with master; so I'm not sure what is going on there.

georgejecook commented:
@TwitchBronBron you waiting on anything else from me? keen to get this in.

TwitchBronBron commented:
I think the code is good. However, we need to add a bunch of documentation examples before this can be merged. These PRs serve as part of the approval process for a new language feature. I'll add a few of them this weekend sometime and get feedback from interested parties.

georgejecook commented:
hey @TwitchBronBron I realized you were right, that dotted get needs to be scope safe, as we will never know for sure if the values are really there, as the developer will expect the alternate and onsequent to be mutually exclusive, and may therefore use values in the either, which could be invalid at runtime. I've updated the code. 

I've also added the documentation.

I hope it's on the right track. I need to add a couple more test cases, to ensure that dotted gets are covered by scope; but I think if you are happy with the

TwitchBronBron commented on bslib.brs:
```diff
@@ -21,4 +21,16 @@ function bslib_toString(value)
+' Simple ternary function. Given a condition, return the
+' consequent if true, and the alternate if false
+'
+function bslib_iff(condition, consequent, alternate)
```
I called this `iff` instead of `ternary` just to save a few characters. If you have a large project with a bunch of source code, `ternary` would start to add up. Plus, `iff` is a [known visual basic function](https://docs.microsoft.com/en-us/office/vba/language/reference/user-interface-help/iif-function) so it still makes sense (at least to me). Thoughts?

georgejecook replied:
I like.

TwitchBronBron commented on src/lexer/TokenKind.ts:
```diff
@@ -73,7 +73,7 @@ export enum TokenKind {
     At = 'At', // @
     Callfunc = 'Callfunc', // @.
-
+    Question = 'Question', // ?
```
I called this `Question` instead of `QuestionMark` because I'm going to name the `??` token `QuestionQuestion` for null coalescing. It felt silly to have `QuestionMark` and `QuestionQuestion`. And I don't like `QuestionMarkQuestionMark`.

Thoughts?

georgejecook replied:
Agree with this too.

TwitchBronBron commented:
Whew, that merge from master was rough! 

@elsassph this is ready for your review.

elsassph commented on src/parser/Expression.ts:
```diff
@@ -1290,6 +1290,90 @@ export class AnnotationExpression extends Expression {
+
+        if (mutatingExpressions.length > 0) {
+            result.push(
+                `(function(condition, `,
```
I think using `condition` for the arg name is risky. Maybe `bslib_cond` will be safer?

TwitchBronBron replied:
Very good point. Will address

TwitchBronBron replied:
I went with `__bsCondition`

elsassph commented on src/parser/Parser.ts:
```diff
@@ -1187,6 +1188,36 @@ export class Parser {
+        if (!test) {
+            test = this.expression();
+        }
+        const questionMarkToken = this.advance();
```
Is/should a newline allowed here?
Maybe some wrapping at places can be safely allowed - and if yes/no it might be good to document.

TwitchBronBron replied:
Good point. I don't see any reason to prevent newlines within ternary statements. In brs, you can use the question mark in place of print, so it might be good to require the condition and the question mark to be in the same line. Thoughts?

The rest can definitely be flexible to use newlines at will. 

So something like this would be supported. 

```
someVar = true = true ?
                      getTrueValue():

                       getFalseValue()
```

TwitchBronBron replied:
Any other expression brine already supported by brs elsewhere would be automatically supported too

elsassph replied:
Basically yes, `?` can't be on a newline but otherwise the rest could be wrapped.

TwitchBronBron replied:
Done. I also added support for mixing comments throughout (but they aren't included in the transpiled output because there's not much benefit in retaining them IMO. (Let me know if you disagree). 

Here are the tests for the various ways you can use 

```
    it('supports multi-line and comments', () => {
        expectZeroDiagnostics(parseBs(`result = true ? \n"zombie"\n: \ntrue`));
        expectZeroDiagnostics(parseBs(`result = true ? "zombie"\n: \ntrue`));
        expectZeroDiagnostics(parseBs(`result = true ? \n"zombie": \ntrue`));
        expectZeroDiagnostics(parseBs(`result = true ? \n"zombie"\n: true`));
        expectZeroDiagnostics(parseBs(`result = true ? "zombie"\n: \ntrue`));
        expectZeroDiagnostics(parseBs(`result = true ? "zombie": \ntrue`));
        expectZeroDiagnostics(parseBs(`result = true ? \n\n\n"zombie": \n\n\n\ntrue`));
        //with comments
        expectZeroDiagnostics(parseBs(`result = true ?'comment\n"zombie"'comment\n:'comment\nntrue`));
    });
    ```

georgejecook replied:
Cool

elsassph commented on src/parser/Parser.ts:
```diff
@@ -1187,6 +1188,36 @@ export class Parser {
+        const alternate = this.expression();
+
+        if (!consequent || !alternate) {
+            this.diagnostics.push({
```
Looks like a malformed ternary will result into something that we'll still be able to transpile - will it work?

TwitchBronBron replied:
Probably not. However, I've tried very hard to still produce partial AST and as much AST recovery as possible, which helps with language server stuff. 

What should malformed transpiled ternary statement look like? Throw an error? Default to a no-op? Write an intentional syntax error so the code doesn't pass the on-device compile phase?

I've kinda just been letting the transpiler write whatever.

elsassph commented on src/DiagnosticMessages.ts:
```diff
@@ -605,6 +605,11 @@ export let DiagnosticMessages = {
         code: 1118,
         severity: DiagnosticSeverity.Error
+    }),
+    malformedTernaryExpression: () => ({
```
Isn't there an unused code you could pick?

TwitchBronBron replied:
Oh yeah, we had one left (1109). I switched to that.

elsassph commented on src/parser/Expression.ts:
```diff
@@ -1290,6 +1290,90 @@ export class AnnotationExpression extends Expression {
+
+    transpile(state: TranspileState) {
+        let result = [];
+        let consequentInfo = util.getExpressionInfo(this.consequent);
```
Does all this code work if `consequent` or `alternate` is `null`? It can happen for malformed expression parsed.

TwitchBronBron replied:
No, I think it will throw a node runtime error because we call `.walk` on the expression. How do you think it should be handled?

TwitchBronBron replied:
I updated this logic so that it will not cause a runtime error during transpile. a null consequent or alternate will be transpiled as `invalid`.

elsassph commented on docs/ternary-operator.md:
````diff
@@ -0,0 +1,165 @@
+a = user = invalid ? "no user" : "logged in"
+```
+
+transpiles to:
````
I think these explanations are a bit confusing.

What about going with:
1. ternary operator with literal values: transpiles with `bslib_ternary`
2. ternary operator with references/function calls: transpiles into a self-executing function with scope capture

TwitchBronBron replied:
Which parts are "thought process" like? I wanted to be transparent in the way the transpiler works, and why we chose the implementations that we did. If there are specific sections that are too verbose, point them out and I'll see if there's a way to trim them down. but I did a quick read-through and still thought most of the information was generally relevant to someone reading the docs.

elsassph commented on docs/ternary-operator.md:
````diff
@@ -0,0 +1,165 @@
+## No standalone statements
+The BrighterScript ternary operator differs from languages like C# and JavaScript in that you cannot use ternary expressions as standalone expressions. This is due to the fact that BrightScript uses `=` for both assignment and equality. As such, ternary expressions must always be part of an assignment or method call. Take a look at the following expression:
+```
+a = myValue ? "a" : "b"
````
I think we want to say that in a simpler way like: 
```
' this is not a valid statement
condition ? doSomething() : orElse()
```
Due to Brightscript syntax where an assignment like `a = b` can not be differentiated from a condition `if a = b`.

elsassph replied:
BTW wouldn't that be ok?
```
' expression hanging alone
(condition ? doSomething() : orElse())
```
